ejb3/src/main/java/org/jboss/as/ejb3/timerservice/TimerServiceImpl.java

```diff
@@ -1180,6 +1211,63 @@ public void afterCompletion(final int status) {
         }
     }
 
+    private class TaskPostPersist extends java.util.TimerTask {
+        private final TimerImpl timer;
+        private int triesCounter = 0;
+
+        TaskPostPersist(TimerImpl timer) {
+            this.timer = timer;
+        }
+
+        TaskPostPersist(TimerImpl timer, int triesCounter) {
+            this.timer = timer;
+            this.triesCounter = triesCounter;
+        }
+
+        @Override
+        public void run() {
+            final ExecutorService executor = executorServiceInjectedValue.getOptionalValue();
+            if (executor != null) {
+                executor.submit(this::persistTimer);
+            }
+        }
+
+        void persistTimer() {
+            final ContextTransactionManager transactionManager = ContextTransactionManager.getInstance();
+            try {
+                transactionManager.begin();
+                timerPersistence.getValue().persistTimer(timer);
+                transactionManager.commit();
+            } catch (Exception e) {
+                try {
+                    transactionManager.rollback();
+                } catch (Exception ee) {
+                    // omit;
+                }
+                EJB3_TIMER_LOGGER.exceptionRunningTimerTask(timer, timer.getTimedObjectId(), e);
+                Date nextExpiration = timer.getNextExpiration();
+                if( nextExpiration != null ){
+                    int max = 10;
+                    String pMax = System.getProperty("jboss.timer.TaskPostPersist.maxRetry");
+                    if( pMax != null ) {
+                        try {
+                            max = Integer.parseInt(pMax);
+                        } catch (NumberFormatException nfe ){
+                            // omit
+                        }
+                    }
+                    if( triesCounter++ < max ) {
+                        long nextTryDelay = nextExpiration.getTime() - System.currentTimeMillis() / 5;
```

Comment:
I don't really like this approach. Isn't this just to deal with transaction serialisation issues, rather than general DB problems? I also don't really understand the logic behind capping it at 5s but only if it is over 1m?

Reply:
Of course, there could be any other DB problems and not only TX serialization. But even in such situations, we have to make this changes in DB. Otherwise, Timer will be dead forever (or it should be recreated). 
I've changed algorithm a little bit. These numbers were just some heuristic from my experiments.
The main idea is to prevent too often repetitions or does not allow to leave it in stale for too long (1 minute).